docs/adr/0007-seperate-binding-generators.md

```diff
@@ -0,0 +1,136 @@
+# Separate the binding generators into their own crates per target
+
+* Status: proposed
+* Deciders: TBD
+* Date: 2022-01-16
+
+Technical Story: [Issue 299](https://github.com/mozilla/uniffi-rs/issues/299)
+
+outdated early prototype: [PR 997](https://github.com/mozilla/uniffi-rs/pull/997)
+
+## Context and Problem Statement
+All the binding generators currently live in the [`uniffi_bindgen`](../../uniffi_bindgen/src/bindings) crate. This creates the following difficulties:
+
+- All the bindings live in the `uniffi` repository, so the `uniffi` team has to maintain them (or at the very least review changes to them).
+- Any changes to a specific binding generator require a new `uniffi_bindgen` release for it to be released. Even if it doesn't impact any of the other bindings.
+- Some bindings require complex build systems to test, and including those build systems in `uniffi` adds its complexity. For example, any type of `gecko-js` bindings would require the mozilla-central build system to build and test.
+- We currently run all the tests for the bindings in our CI and through `cargo test`. This means that if one binding target gets outdated and fails, or if a developer doesn't have the needed libraries installed for one of the targets, the tests would fail.
+- It's currently difficult to support third-party developers writing bindings for languages the core team does not care about.
+
+This ADR re-evaluates the architecture of how we consume the bindings in hope of improving the developer experience of adding new bindings, testing new and the current bindings and simplifying the ownership of the bindings.
+## Decision Drivers
+
+* Maintainability, we would like to improve the maintainability of our binding generators - this means clear ownership of the bindings.
+* Testability, it should be easy for developers to test the bindings they care about. Without having to navigate and install unfamiliar libraries and build systems.
+* Version compatibility, this refers to the same version of `uniffi` being used to generate both the scaffolding and the bindings. This includes:
+    - How easy is it to accidentally have a version mismatch
+    - How much damage is caused by the the version mismatch
+    - The difficulty of debugging a version mismatch
+* Developer experience, it should be easier to write and maintain a new binding generator than it currently is.
```

Comment:
my very naive thought is that it could be easier for third-party developers to write their own bindings without having it live in this repository could make it easier... but I don't have anything concrete to back that claim 🙃